Use tuples in APIs that return multiple values instead of out parameters

tmat opened the pull request.

tmat commented:
@nguerrera 
@gafter @jcouv @MadsTorgersen FYI
@terrajobst @KrzysztofCwalina heads up: Tuples coming to public API

jcouv commented on src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/BlobEncoders.cs:
```diff
@@ -74,24 +75,21 @@ public MethodSignatureEncoder PropertySignature(bool isInstanceProperty = false)
         }
 
-        public void CustomAttributeSignature(out FixedArgumentsEncoder fixedArguments, out CustomAttributeNamedArgumentsEncoder namedArguments)
+        [return: TupleElementNames(new[] { "FixedArguments", "NamedArguments" })]
```
This seems correct, but be aware that the compiler does not generate such attribute (yet). @VSadov will work on that the week after next. 
If the serialization format needs to change for some compiler-motivated reason, this would have to be updated.

Aside, from that, I'm excited to see our first ValueTuple consumer! ;-)

VSadov replied:
So far it looks correct. 
For tuples with length under 8 it is very likely that this will be the encoding.

tmat replied:
Yes, changes are expected. This will likely ship with Dev15, so I hope we'll switch to real tuple syntax at some point before Dev15 RTM.

gafter replied:
There has been some recent discussion of changing the encoding. @agocke Are we prepared to commit permanently to this particular encoding?

tmat replied:
@gafter The library is not shipping until Dev15, so changes are ok.

agocke replied:
For the limited use case @tmat is using here, I think it's very solid. I'm not concerned about changing this, since I assume we'll eventually want S.R.M to be compiled with a 2.0.0 compiler, so we'll handle the encoding in the compiler.

nguerrera commented:
:+1:

I'm OK with the direction, but I will note that readability is temporarily reduced at call sites while we wait for language support.

Is there are a committed timeline for language support to land?

Does corefx have a plan for picking up new compiler releases soon after they ship. (We waited quite a while for C# 6 and I'm hoping we'll be getting >6 much sooner after they ship.)

stephentoub commented:
FWIW, I'm still hesitant to see tuples used like this in public APIs, especially in corefx public APIs, and would prefer to see struct types defined, that can be documented, that work by name with reflection and dynamic, that work with any .NET language and not just those that (eventually) recognize the name attributes, that don't necessitate an ecosystem of tools be updated before they can be understood (e.g. decompilers, debuggers), etc.

tmat commented:
Considering all the feedback above I have decided to use a hybrid approach and updated the PR.

It makes sense to avoid tuples for ReserveXxx APIs and declare a struct. 

For signature encoders I believe using anonymous tuples is the best approach. It is absolutely critical for correct usage of the APIs that we convey the order in which the items of the tuple must be used in order to encode the signature properly. If we define a custom structs with properties whose names correspond to the names of the encoders we lose the ordering information (what's worse, completion sorts the members alphabetically and thus the order in which the encoders are expected to be used is completely lost). Using tuples with unnamed items forces the users to use Item1, Item2, etc. which is **desirable**, unlike other common uses of tuples, as it communicates the order. Besides naming the items doesn't add much value since the types of the Item* properties already convey the same information.

In C# 7 one would use deconstruction syntax anyways, for example:

```C#
var (argumentType, name, literal) = namedArguments.AddArgument(isField: false);
```

To emphasize the importance of order I have added doc comments to each method with explaination.

nguerrera commented on src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/BlobEncoders.cs:
```diff
@@ -149,7 +205,20 @@ public MethodSignatureEncoder(BlobBuilder builder, bool hasVarArgs)
+        /// Returns a pair of encoders that must be used in the order they appear in the tuple:
+        /// first <see cref="ReturnTypeEncoder"/> to encode the return type,
+        /// then <see cref="ParametersEncoder"/> to encode the actual parameters.
+        /// </returns>
```
Do we need to duplicate this text? From offline discussion, the issue is that returns docs don't show up in IntelliSense and this is important to read. I would just strip the repetitive returns doc.

tmat replied:
I guess we don't need the ```<returns>```. I'll remove them.

nguerrera commented:
:+1:

VSadov commented:
LGTM

A very interesting insight on how tuples might or might not be used in public APIs in general.

tmat commented:
CentOS failure: https://github.com/dotnet/corefx/issues/10375
